File: sourcecode/Android/dConnectDevicePlugin/dConnectDeviceTest/src/com/nttdocomo/android/dconnect/deviceplugin/test/profile/TestDeviceOrientationProfile.java
/*
 TestDeviceOrientationProfile.java
 Copyright (c) 2014 NTT DOCOMO,INC.
 Released under the MIT license
 http://opensource.org/licenses/mit-license.php
 */
package com.nttdocomo.android.dconnect.deviceplugin.test.profile;

import android.content.Intent;
import android.os.Bundle;

import com.nttdocomo.android.dconnect.message.MessageUtils;
import com.nttdocomo.android.dconnect.profile.DeviceOrientationProfile;
import com.nttdocomo.dconnect.message.DConnectMessage;

/**
 * JUnit用テストデバイスプラグイン、Connectプロファイル.
 * 
 * @author NTT DOCOMO, INC.
 */
public class TestDeviceOrientationProfile extends DeviceOrientationProfile {

    /**
     * デバイスIDをチェックする.
     * @param deviceId デバイスID
     * @return <code>deviceId</code>がテスト用デバイスIDに等しい場合はtrue、そうでない場合はfalse
     */
    private boolean checkDeviceId(final String deviceId) {
        return TestNetworkServiceDiscoveryProfile.DEVICE_ID.equals(deviceId);
    }

    /**
     * デバイスIDが空の場合のエラーを作成する.
     * @param response レスポンスを格納するIntent
     */
    private void createEmptyDeviceId(final Intent response) {
        setResult(response, DConnectMessage.RESULT_ERROR);
    }

    /**
     * デバイスが発見できなかった場合のエラーを作成する.
     * @param response レスポンスを格納するIntent
     */
    private void createNotFoundDevice(final Intent response) {
        setResult(response, DConnectMessage.RESULT_ERROR);
    }

    @Override
    protected boolean onPutOnDeviceOrientation(final Intent request, final Intent response, final String deviceId, 
            final String sessionKey) {
        if (deviceId == null) {
            createEmptyDeviceId(response);
        } else if (!checkDeviceId(deviceId)) {
            createNotFoundDevice(response);
        } else if (sessionKey == null) {
            MessageUtils.setInvalidRequestParameterError(response);
        } else {
            setResult(response, DConnectMessage.RESULT_OK);

            Intent intent = MessageUtils.createEventIntent();
            setSessionKey(intent, sessionKey);
            setDeviceID(intent, deviceId);
            setProfile(intent, getProfileName());
            setAttribute(intent, ATTRIBUTE_ON_DEVICE_ORIENTATION);
            
            Bundle orientation = new Bundle();
            Bundle a1 = new Bundle();
            setX(a1, 0.0);
            setY(a1, 0.0);
            setZ(a1, 0.0);
            Bundle a2 = new Bundle();
            setX(a2, 0.0);
            setY(a2, 0.0);
            setZ(a2, 0.0);
            Bundle r = new Bundle();
            setAlpha(r, 0.0);
            setBeta(r, 0.0);
            setGamma(r, 0.0);
            
            setAcceleration(orientation, a1);
            setAccelerationIncludingGravity(orientation, a2);
            setRotationRate(orientation, r);
            setInterval(orientation, 0);
            
            setOrientation(intent, orientation);
            Util.sendBroadcast(getContext(), intent);
        }
        return true;
    }

    @Override
    protected boolean onDeleteOnDeviceOrientation(final Intent request, final Intent response, final String deviceId, 
            final String sessionKey) {
        if (deviceId == null) {
            createEmptyDeviceId(response);
        } else if (!checkDeviceId(deviceId)) {
            createNotFoundDevice(response);
        } else if (sessionKey == null) {
            MessageUtils.setInvalidRequestParameterError(response);
        } else {
            setResult(response, DConnectMessage.RESULT_OK);
        }
        return true;
    }

}

File: sourcecode/iOS/dConnectDevicePlugin/dConnectDeviceTest/dConnectDeviceTest/Classes/Profile/TestDeviceOrientationProfile.m
//
//  TestDeviceOrientationProfile.h
//  DConnectSDK
//
//  Copyright (c) 2014 NTT DOCOMO, INC.
//  Released under the MIT license
//  http://opensource.org/licenses/mit-license.php
//

#import "TestDeviceOrientationProfile.h"
#import "DeviceTestPlugin.h"

@implementation TestDeviceOrientationProfile

- (id) initWithDevicePlugin:(DeviceTestPlugin *)plugin {
    self = [super init];
    
    if (self) {
        self.delegate = self;
        _plugin = plugin;
    }
    
    return self;
}


#pragma mark - Put Methods
#pragma mark Event Registration

- (BOOL) profile:(DConnectDeviceOrientationProfile *)profile didReceivePutOnDeviceOrientationRequest:(DConnectRequestMessage *)request
        response:(DConnectResponseMessage *)response deviceId:(NSString *)deviceId sessionKey:(NSString *)sessionKey
{
    
    CheckDIDAndSK(response, deviceId, sessionKey) {
        response.result = DConnectMessageResultTypeOk;
        
        DConnectMessage *event = [DConnectMessage message];
        [event setString:sessionKey forKey:DConnectMessageSessionKey];
        [event setString:deviceId forKey:DConnectMessageDeviceId];
        [event setString:self.profileName forKey:DConnectMessageProfile];
        [event setString:DConnectDeviceOrientationProfileName
                  forKey:DConnectMessageAttribute];
        
        DConnectMessage *orientation = [DConnectMessage message];
        
        DConnectMessage *a1 = [DConnectMessage message];
        [DConnectDeviceOrientationProfile setX:0 target:a1];
        [DConnectDeviceOrientationProfile setY:0 target:a1];
        [DConnectDeviceOrientationProfile setZ:0 target:a1];
        
        DConnectMessage *a2 = [DConnectMessage message];
        [DConnectDeviceOrientationProfile setX:0 target:a2];
        [DConnectDeviceOrientationProfile setY:0 target:a2];
        [DConnectDeviceOrientationProfile setZ:0 target:a2];

        DConnectMessage *r = [DConnectMessage message];
        [DConnectDeviceOrientationProfile setAlpha:0 target:r];
        [DConnectDeviceOrientationProfile setBeta:0 target:r];
        [DConnectDeviceOrientationProfile setGamma:0 target:r];
        
        [DConnectDeviceOrientationProfile setAcceleration:a1 target:orientation];
        [DConnectDeviceOrientationProfile setAccelerationIncludingGravity:a2 target:orientation];
        [DConnectDeviceOrientationProfile setRotationRate:r target:orientation];
        [DConnectDeviceOrientationProfile setInterval:0 target:orientation];
        
        [DConnectDeviceOrientationProfile setOrientation:orientation target:event];
        [_plugin asyncSendEvent:event];
    }
    
    return YES;
}

#pragma mark - Delete Methods
#pragma mark Event Unregistration

- (BOOL) profile:(DConnectDeviceOrientationProfile *)profile didReceiveDeleteOnDeviceOrientationRequest:(DConnectRequestMessage *)request
        response:(DConnectResponseMessage *)response deviceId:(NSString *)deviceId sessionKey:(NSString *)sessionKey
{
    
    CheckDIDAndSK(response, deviceId, sessionKey) {
        response.result = DConnectMessageResultTypeOk;
    }
    
    return YES;
}


@end
